witness/golang/cmd/witness/internal/witness/witness_test.go

```diff
@@ -0,0 +1,287 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"context"
+	"crypto/rand"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/google/trillian/merkle/logverifier"
+	"github.com/google/trillian/merkle/rfc6962/hasher"
+	_ "github.com/mattn/go-sqlite3" // Load drivers for sqlite3
+	"golang.org/x/mod/sumdb/note"
+)
+
+var (
+	initChkpt = Chkpt{
+		Size: 5,
+		Raw:  []byte("Log Checkpoint v0\n5\n41smjBUiAU70EtKlT6lIOIYtRTYxYXsDB+XHfcvu/BE=\n"),
+	}
+	newChkpt = Chkpt{
+		Size: 8,
+		Raw:  []byte("Log Checkpoint v0\n8\nV8K9aklZ4EPB+RMOk1/8VsJUdFZR77GDtZUQq84vSbo=\n"),
+	}
+	consProof = [][]byte{
+		dh("b9e1d62618f7fee8034e4c5010f727ab24d8e4705cb296c374bf2025a87a10d2", 32),
+		dh("aac66cd7a79ce4012d80762fe8eec3a77f22d1ca4145c3f4cee022e7efcd599d", 32),
+		dh("89d0f753f66a290c483b39cd5e9eafb12021293395fad3d4a2ad053cfbcfdc9e", 32),
+		dh("29e40bb79c966f4c6fe96aff6f30acfce5f3e8d84c02215175d6e018a5dee833", 32),
+	}
+)
+
+func signChkpts(skey string, chkpts []string) ([][]byte, error) {
+	ns, err := note.NewSigner(skey)
+	if err != nil {
+		return nil, fmt.Errorf("signChkpt: couldn't create signer")
+	}
+	signed := make([][]byte, len(chkpts))
+	for i, c := range chkpts {
+		s, err := note.Sign(&note.Note{Text: c}, ns)
+		if err != nil {
+			return nil, fmt.Errorf("signChkpt: couldn't sign note")
+		}
+		signed[i] = s
+	}
+	return signed, nil
+}
+
+func newOpts(d *sql.DB, logID string, logPK string, wSK string) (Opts, error) {
+	ns, err := note.NewSigner(wSK)
+	if err != nil {
+		return Opts{}, fmt.Errorf("setOpts: couldn't create a note signer")
+	}
+	h := hasher.DefaultHasher
+	logV, err := note.NewVerifier(logPK)
+	if err != nil {
+		return Opts{}, fmt.Errorf("setOpts: couldn't create a log verifier")
+	}
+	sigVs := []note.Verifier{logV}
+	log := LogInfo{
+		SigVs: sigVs,
+		LogV:  logverifier.New(h),
+	}
+	m := map[string]LogInfo{logID: log}
+	opts := Opts{
+		DB:        d,
+		Signer:    ns,
+		KnownLogs: m,
+	}
+	return opts, nil
+}
+
+func newOptsAndKeys(d *sql.DB, logID string, logPK string) (Opts, error) {
+	wSK, _, err := note.GenerateKey(rand.Reader, "witness")
+	if err != nil {
+		return Opts{}, fmt.Errorf("couldn't generate witness keys")
+	}
+	return newOpts(d, logID, logPK, wSK)
+}
+
+func newWitness(t *testing.T, d *sql.DB, logID string, logPK string) *Witness {
+	opts, err := newOptsAndKeys(d, logID, logPK)
+	if err != nil {
+		t.Fatalf("couldn't create witness opt struct: %v", err)
+	}
+	w, err := New(opts)
+	if err != nil {
+		t.Fatalf("couldn't create witness: %v", err)
+	}
+	return w
+}
+
+// dh is taken from https://github.com/google/trillian/blob/master/merkle/logverifier/log_verifier_test.go.
+func dh(h string, expLen int) []byte {
```

Comment:
nit: you could pass in the `t` here, call `t.Helper()` on the first line, and then replace the `panic` with `t.Fatalf()` 

(advantages are that you won't abort other tests running in the same process, you'll get a nice clean error without stack, and thanks to `t.Helper` the failure will be associated with the line of the test which called this func)

Reply:
I don't actually see a way to do this, since `dh` is used to define the `consProof` var, which is done outside of any test/function?